streams/src/test/java/org/apache/kafka/streams/integration/GlobalThreadShutDownOrderTest.java

```diff
@@ -0,0 +1,217 @@
+                               Consumed.with(Serdes.String(), Serdes.Long()),
+                               new MockProcessorSupplier());
+
+        stream = builder.stream(streamTopic, stringLongConsumed);
+
+        stream.process(new ProcessorSupplier<String, Long>() {
+            @Override
+            public Processor<String, Long> get() {
+                return new GlobalStoreProcessor(globalStore);
+            }
+        });
+
+    }
+
+    @After
+    public void whenShuttingDown() throws IOException {
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+
+    @Test
+    public void shouldFinishGlobalStoreOperationOnShutDown() throws Exception {
+        kafkaStreams = new KafkaStreams(builder.build(), streamsConfiguration);
+        populateTopics(globalStoreTopic);
+        populateTopics(streamTopic);
+
+        kafkaStreams.start();
+
+        TestUtils.waitForCondition(new TestCondition() {
+            @Override
+            public boolean conditionMet() {
+                return retrievedValuesList.size() > 0;
+            }
+        }, 5000L, "Did not retrieve any values");
+
+        kafkaStreams.close(30, TimeUnit.SECONDS);
+
+        List<Long> expectedRetrievedValues = Arrays.asList(1L, 2L, 3L, 4L);
+        assertEquals(expectedRetrievedValues, retrievedValuesList);
+    }
+
+
+    private void createTopics() throws InterruptedException {
+        streamTopic = "stream-topic";
+        globalStoreTopic = "global-store-topic";
+        CLUSTER.createTopics(streamTopic);
+        CLUSTER.createTopic(globalStoreTopic);
+    }
+
+
+    private void populateTopics(String topicName) throws Exception {
+        IntegrationTestUtils.produceKeyValuesSynchronously(
+            topicName,
+            Arrays.asList(
+                new KeyValue<>("A", 1L),
+                new KeyValue<>("B", 2L),
+                new KeyValue<>("C", 3L),
+                new KeyValue<>("D", 4L)),
+            TestUtils.producerConfig(
+                CLUSTER.bootstrapServers(),
+                StringSerializer.class,
+                LongSerializer.class,
+                new Properties()),
+            mockTime);
+    }
+
+
+    private class GlobalStoreProcessor extends AbstractProcessor<String, Long> {
+
+        private KeyValueStore<String, Long> store;
+        private final String storeName;
+
+        GlobalStoreProcessor(final String storeName) {
+
+            this.storeName = storeName;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public void init(final ProcessorContext context) {
+            super.init(context);
+            store = (KeyValueStore<String, Long>) context.getStateStore(storeName);
+        }
+
+        @Override
+        public void process(final String key, final Long value) {
+            // need to simulate slow processing
+            Utils.sleep(1000);
```

Comment:
I think a better test scenario is to move the logic in `close()` call, i.e. when the stream thread is being shutdown, and topology is closing, we call `processorNode.close()` in which we wait for a while and then tries to access the global store. It mimics the case where in closing the store cache is flushed and hence tries to access the global store again.

Reply:
ack, will move